### PR TITLE
Fix database PANIC issue for "grant all on all tables in schema xxx to yyy;"

### PR DESCRIPTION
**Pull request by ginobiliwang**
1) We have found in gpdb 6 versions there is an issue when using  "grant all on all tables in schema xxx to yyy;",  tables with partitions or inherits child tables, will lead to database PANIC.

2) After analyze coredump informations, we have found that the data structure that used to decode objects on segments has been incorrectly used, which leads to NullPointer and PANIC.

Screen Shot 2019-09-10 at 10 23 53 PM

3) After analyzing code logical, we have found that there is flag call "added_objs", which will be used in the master node to encode objects that have partition references into "RangerVar" format. It is the root cause of the issue. 
This PR has used NodeType information to tell different kinds of object and fix the issue.

4) Will add some test cases also in this PR.

### CODE REVIEW

**Comment by ginobiliwang** [comment]
Some more finding for this issue. 
I have checked the commit history for each piece of code in this file (aclchk.c).  
The encoding part for list object is primarily committed by these two commits,  
https://github.com/greenplum-db/gpdb/commit/e5d17790c185217831828169884f992be32502a6
https://github.com/greenplum-db/gpdb/commit/6b0e52beadd678c5050af9c978c26d171ba86ae0

These are very old commits, the second one is the initial commit for gpdb. 
In other words, i would assume that test cases(partition and inherit table) for this issue has never been covered.

**Comment by gkokolatos** [comment]
> These are very old commits, the second one is the initial commit for gpdb.
> In other words, i would assume that test cases(partition and inherit table) for this issue has never been covered.

There are indeed. I am still going to argue for removing the RangerVar entries. I guess that part of code might have made sense when the catalog was not present on the segments but I do not see a reason for it now. Of course you know best and I might be wrong. 

Any progress with the test cases? May I help with those somehow?

**Comment by ginobiliwang** [comment]
> There are indeed. I am still going to argue for removing the RangerVar entries. I guess that part of code might have made sense when the catalog was not present on the segments but I do not see a reason for it now. Of course you know best and I might be wrong.

I assume that  for these code pieces,  the origin idea of the author is to do caching work. He want to send bosh "nspname" and "relname" with RangerVar struct to segments directly,  and there may have specific function to get them, which could prevent from using ordinary ones from time to time.

https://github.com/greenplum-db/gpdb/blob/66199b3d96ac99c60d7b5845ca0900196c148409/src/backend/catalog/aclchk.c#L473-L489

Meanwhile, from code logic, i see that only "nspname" has been used (correct me if i am wrong). Its indeed more efficient to use "strVal".  Also, i agree to get rid of catalog things in segment as you said.

From git history, it shows that these are from "Initial Greenplum code dump", so if possible i could have a try to verify if it is Ok to remove RangerVar in these piece of code logic.

What is your thought?

> 
> Any progress with the test cases? May I help with those somehow?
sorry, a little busy these days, let me make it up in this weekend.

**Comment by gkokolatos** [comment]
> https://github.com/greenplum-db/gpdb/blob/66199b3d96ac99c60d7b5845ca0900196c148409/src/backend/catalog/aclchk.c#L473-L489
> 
> Meanwhile, from code logic, i see that only "nspname" has been used (correct me if i am wrong). Its indeed more efficient to use "strVal". Also, i agree to get rid of catalog things in segment as you said.
> 
> From git history, it shows that these are from "Initial Greenplum code dump", so if possible i could have a try to verify if it is Ok to remove RangerVar in these piece of code logic.
> 
> What is your thought?

I think you are spot on! I would argue that the whole block can be removed and it should address the issue. Also that way we have less deviations with upstream. Up to you.

I am more curious with the test cases. I have been able to reproduce given your initial report, though I am not certain if I am missing something.

> sorry, a little busy these days, let me make it up in this weekend.

No need to apologise nor make anything up during the weekend! We can pick it up during the week!

**Comment by ginobiliwang** [comment]
> > https://github.com/greenplum-db/gpdb/blob/66199b3d96ac99c60d7b5845ca0900196c148409/src/backend/catalog/aclchk.c#L473-L489
> > 
> > Meanwhile, from code logic, i see that only "nspname" has been used (correct me if i am wrong). Its indeed more efficient to use "strVal". Also, i agree to get rid of catalog things in segment as you said.
> > From git history, it shows that these are from "Initial Greenplum code dump", so if possible i could have a try to verify if it is Ok to remove RangerVar in these piece of code logic.
> > What is your thought?
> 
> I think you are spot on! I would argue that the whole block can be removed and it should address the issue. Also that way we have less deviations with upstream. Up to you.
> 

Great! That is actually my original idea to resolve the issue(to remove this code block). Since i am not a committer for gpdb, things would be sort of challenging if i would like to delete some code but add/update them,  so i choose to add something.
Let me draft a commit for this.

> I am more curious with the test cases. I have been able to reproduce given your initial report, though I am not certain if I am missing something.
> 

Will add test cases in the commit mentioned above (simple table, partition and inherit table, mix of them).

**Comment by ginobiliwang** [comment]
> Great work! I took the liberty to update the test results so that the CI builds green.
> 
> It needs squashing and back porting to 6X_STABLE which I will gladly take care off once the CI run is completed.


Thanks a lot. Looks like the CI has some issue to complete for the recent two commits. Is it broken?  :(

**Comment by gkokolatos** [comment]
> > Great work! I took the liberty to update the test results so that the CI builds green.
> > It needs squashing and back porting to 6X_STABLE which I will gladly take care off once the CI run is completed.
> 
> Thanks a lot. Looks like the CI has some issue to complete for the recent two commits. Is it broken? :(

The failures seem completely unrelated to this PR. Of course that is not a reason to not get another try. I will rebase the branch and force push to trigger another test run.

**Comment by berlin-ab** [comment]
This, unfortunately, does cause the build to fail.

**Comment by berlin-ab** [comment]
There's a gpcheckcat failure that is difficult to notice at the end of ICW.

**Comment by berlin-ab** [comment]
Screen Shot 2019-09-24 at 4 07 53 PM

**Comment by edespino** [comment]
I was able to hop into the CI failed job container and reproduce the gpcheckcat issue (I previously started the gpdemo cluster). I believe gpcheckcat is executed for all databases after the installcheck-world run. And as Adam indicated, the error is somewhat hidden amongst the volumnous output:

```
fly --target gpdb-prod intercept -u <job id>
1: build #724, step: bin_gpdb, type: get
2: build #724, step: ic_gpdb, type: task
choose a container: 2
[root@f4ed187b-17b2-46f0-7f19-e6656cf3eaad e18b2f02]# su - gpadmin
[gpadmin@f4ed187b-17b2-46f0-7f19-e6656cf3eaad ~]$ source /tmp/build/e18b2f02/gpdb_src/gpAux/gpdemo/gpdemo-env.sh
[gpadmin@f4ed187b-17b2-46f0-7f19-e6656cf3eaad ~]$ source /usr/local/greenplum-db-devel/greenplum_path.sh
[gpadmin@f4ed187b-17b2-46f0-7f19-e6656cf3eaad ~]$ gpcheckcat regression
Truncated batch size to number of primaries: 4

Connected as user 'gpadmin' to database 'regression', port '15432', gpdb version '6.0'
-------------------------------------------------------------------
Batch size: 4
Performing test 'unique_index_violation'
Total runtime for test 'unique_index_violation': 0:00:02.32
Performing test 'duplicate'
Total runtime for test 'duplicate': 0:00:03.01
Performing test 'missing_extraneous'
Total runtime for test 'missing_extraneous': 0:00:02.62
Performing test 'inconsistent'
Total runtime for test 'inconsistent': 0:00:08.71
Performing test 'foreign_key'
Total runtime for test 'foreign_key': 0:00:04.48
Performing test 'acl'
Total runtime for test 'acl': 0:00:00.53
Performing test 'pgclass'
Total runtime for test 'pgclass': 0:00:00.11
Performing test 'namespace'
Total runtime for test 'namespace': 0:00:00.02
Performing test 'distribution_policy'
Total runtime for test 'distribution_policy': 0:00:00.04
Performing test 'dependency'
Total runtime for test 'dependency': 0:00:00.74
Performing test 'owner'
Total runtime for test 'owner': 0:00:00.11
Performing test 'part_integrity'
Total runtime for test 'part_integrity': 0:00:03.00
Performing test 'part_constraint'
Total runtime for test 'part_constraint': 0:00:02.15
Performing test 'orphaned_toast_tables'
Total runtime for test 'orphaned_toast_tables': 0:00:00.18

SUMMARY REPORT
===================================================================
Completed 14 test(s) on database 'regression' at 2019-09-24 20:06:52 with elapsed time 0:00:28
Found a total of 1 issue(s)

----------------------------------------------------
Relation oid: 135849
Relation schema: bfv_partition
Relation name: part_acl_test_1_prt_p1

    Name of test which found this issue: acl_pg_class
        Master (f4ed187b-17b2-46f0-7f19-e6656cf3eaad:15432) and content 0 (f4ed187b-17b2-46f0-7f19-e6656cf3eaad:25432) have different ACLs:
            Exist(s) on master only: ['part_acl_u1=r/part_acl_owner', 'part_acl_owner=arwdDxt/part_acl_owner']
        Master (f4ed187b-17b2-46f0-7f19-e6656cf3eaad:15432) and content 1 (f4ed187b-17b2-46f0-7f19-e6656cf3eaad:25433) have different ACLs:
            Exist(s) on master only: ['part_acl_u1=r/part_acl_owner', 'part_acl_owner=arwdDxt/part_acl_owner']
        Master (f4ed187b-17b2-46f0-7f19-e6656cf3eaad:15432) and content 2 (f4ed187b-17b2-46f0-7f19-e6656cf3eaad:25434) have different ACLs:
            Exist(s) on master only: ['part_acl_u1=r/part_acl_owner', 'part_acl_owner=arwdDxt/part_acl_owner']


[gpadmin@f4ed187b-17b2-46f0-7f19-e6656cf3eaad ~]$ logout
```

**Comment by gkokolatos** [comment]
> There's a gpcheckcat failure that is difficult to notice at the end of ICW.

Thank you for this! I am addressing it.

**Comment by ginobiliwang** [comment]
I have reproduced the issue in the CI error reporting.

```
[gpadmin@sdw1 ~]$ psql
psql (9.4.24)
Type "help" for help.

gpadmin=# drop schema if exists bfv_partition cascade;
NOTICE:  schema "bfv_partition" does not exist, skipping
DROP SCHEMA
gpadmin=# DROP USER mpp3641_user;
ERROR:  role "mpp3641_user" does not exist
gpadmin=# DROP ROLE part_acl_owner;
DROP ROLE
gpadmin=# DROP ROLE part_acl_u1;
DROP ROLE
gpadmin=# create schema bfv_partition;
CREATE SCHEMA
gpadmin=# DROP ROLE IF EXISTS part_acl_owner;
NOTICE:  role "part_acl_owner" does not exist, skipping
DROP ROLE
gpadmin=# CREATE ROLE part_acl_owner;
NOTICE:  resource queue required -- using default resource queue "pg_default"
CREATE ROLE
gpadmin=# DROP ROLE IF EXISTS part_acl_u1;
NOTICE:  role "part_acl_u1" does not exist, skipping
DROP ROLE
gpadmin=# CREATE ROLE part_acl_u1;
NOTICE:  resource queue required -- using default resource queue "pg_default"
CREATE ROLE
gpadmin=# GRANT ALL ON SCHEMA bfv_partition to part_acl_owner;
GRANT
gpadmin=# SET ROLE part_acl_owner;
SET
gpadmin=> CREATE TABLE part_acl_test (id int4) PARTITION BY LIST (id) (PARTITION p1 VALUES (1));
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'id' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
NOTICE:  CREATE TABLE will create partition "part_acl_test_1_prt_p1" for table "part_acl_test"
CREATE TABLE
gpadmin=> GRANT SELECT ON part_acl_test TO part_acl_u1;
GRANT
gpadmin=> ALTER TABLE part_acl_test ADD PARTITION p2 VALUES (2);
NOTICE:  CREATE TABLE will create partition "part_acl_test_1_prt_p2" for table "part_acl_test"
ALTER TABLE
gpadmin=> \q
[gpadmin@sdw1 ~]$
[gpadmin@sdw1 ~]$ gpcheckcat gpadmin
Truncated batch size to number of primaries: 2

Connected as user 'gpadmin' to database 'gpadmin', port '5432', gpdb version '6.0'
-------------------------------------------------------------------
Batch size: 2
Performing test 'unique_index_violation'
Total runtime for test 'unique_index_violation': 0:00:00.48
Performing test 'duplicate'
Total runtime for test 'duplicate': 0:00:02.79
Performing test 'missing_extraneous'
Total runtime for test 'missing_extraneous': 0:00:00.53
Performing test 'inconsistent'
Total runtime for test 'inconsistent': 0:00:03.54
Performing test 'foreign_key'
Total runtime for test 'foreign_key': 0:00:00.78
Performing test 'acl'
Total runtime for test 'acl': 0:00:00.05
Performing test 'pgclass'
Total runtime for test 'pgclass': 0:00:00.01
Performing test 'namespace'
Total runtime for test 'namespace': 0:00:00.00
Performing test 'distribution_policy'
Total runtime for test 'distribution_policy': 0:00:00.00
Performing test 'dependency'
Total runtime for test 'dependency': 0:00:00.20
Performing test 'owner'
Total runtime for test 'owner': 0:00:00.06
Performing test 'part_integrity'
Total runtime for test 'part_integrity': 0:00:00.04
Performing test 'part_constraint'
Total runtime for test 'part_constraint': 0:00:00.20
Performing test 'orphaned_toast_tables'
Total runtime for test 'orphaned_toast_tables': 0:00:00.01

SUMMARY REPORT
===================================================================
Completed 14 test(s) on database 'gpadmin' at 2019-09-25 01:38:25 with elapsed time 0:00:08
Found a total of 1 issue(s)

----------------------------------------------------
Relation oid: 24121
Relation schema: public
Relation name: part_acl_test_1_prt_p1

    Name of test which found this issue: acl_pg_class
        Master (sdw1:5432) and content 0 (sdw1:6000) have different ACLs:
            Exist(s) on master only: ['part_acl_u1=r/part_acl_owner', 'part_acl_owner=arwdDxt/part_acl_owner']

```

After i undo the deletion of the very code block, issue is gone.

```
[gpadmin@sdw1 ~]$ psql
psql (9.4.24)
Type "help" for help.

gpadmin=# drop schema if exists bfv_partition cascade;
NOTICE:  schema "bfv_partition" does not exist, skipping
DROP SCHEMA
gpadmin=# DROP USER mpp3641_user;
ERROR:  role "mpp3641_user" does not exist
gpadmin=# DROP ROLE part_acl_u1;
ERROR:  role "part_acl_u1" cannot be dropped because some objects depend on it
DETAIL:  privileges for table part_acl_test_1_prt_p2
privileges for table part_acl_test
privileges for table part_acl_test_1_prt_p1
gpadmin=# drop table part_acl_test;
DROP TABLE
gpadmin=# DROP ROLE part_acl_u1;
DROP ROLE
gpadmin=# DROP ROLE part_acl_owner;
DROP ROLE
gpadmin=# drop schema if exists bfv_partition cascade;
NOTICE:  schema "bfv_partition" does not exist, skipping
DROP SCHEMA
gpadmin=# create schema bfv_partition;
CREATE SCHEMA
gpadmin=# DROP ROLE IF EXISTS part_acl_owner;
NOTICE:  role "part_acl_owner" does not exist, skipping
DROP ROLE
gpadmin=# CREATE ROLE part_acl_owner;
NOTICE:  resource queue required -- using default resource queue "pg_default"
CREATE ROLE
gpadmin=# DROP ROLE IF EXISTS part_acl_u1;
NOTICE:  role "part_acl_u1" does not exist, skipping
DROP ROLE
gpadmin=# CREATE ROLE part_acl_u1;
NOTICE:  resource queue required -- using default resource queue "pg_default"
CREATE ROLE
gpadmin=# GRANT ALL ON SCHEMA bfv_partition to part_acl_owner;
GRANT
gpadmin=# SET ROLE part_acl_owner;
SET
gpadmin=> CREATE TABLE part_acl_test (id int4) PARTITION BY LIST (id) (PARTITION p1 VALUES (1));
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'id' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
NOTICE:  CREATE TABLE will create partition "part_acl_test_1_prt_p1" for table "part_acl_test"
CREATE TABLE
gpadmin=> GRANT SELECT ON part_acl_test TO part_acl_u1;
GRANT
gpadmin=> ALTER TABLE part_acl_test ADD PARTITION p2 VALUES (2);
NOTICE:  CREATE TABLE will create partition "part_acl_test_1_prt_p2" for table "part_acl_test"
ALTER TABLE
gpadmin=> \q
[gpadmin@sdw1 ~]$
[gpadmin@sdw1 ~]$
[gpadmin@sdw1 ~]$
[gpadmin@sdw1 ~]$
[gpadmin@sdw1 ~]$
[gpadmin@sdw1 ~]$ gpcheckcat gpadmin
Truncated batch size to number of primaries: 2

Connected as user 'gpadmin' to database 'gpadmin', port '5432', gpdb version '6.0'
-------------------------------------------------------------------
Batch size: 2
Performing test 'unique_index_violation'
Total runtime for test 'unique_index_violation': 0:00:00.48
Performing test 'duplicate'
Total runtime for test 'duplicate': 0:00:02.82
Performing test 'missing_extraneous'
Total runtime for test 'missing_extraneous': 0:00:00.53
Performing test 'inconsistent'
Total runtime for test 'inconsistent': 0:00:03.69
Performing test 'foreign_key'
Total runtime for test 'foreign_key': 0:00:00.78
Performing test 'acl'
Total runtime for test 'acl': 0:00:00.05
Performing test 'pgclass'
Total runtime for test 'pgclass': 0:00:00.01
Performing test 'namespace'
Total runtime for test 'namespace': 0:00:00.00
Performing test 'distribution_policy'
Total runtime for test 'distribution_policy': 0:00:00.00
Performing test 'dependency'
Total runtime for test 'dependency': 0:00:00.20
Performing test 'owner'
Total runtime for test 'owner': 0:00:00.06
Performing test 'part_integrity'
Total runtime for test 'part_integrity': 0:00:00.04
Performing test 'part_constraint'
Total runtime for test 'part_constraint': 0:00:00.20
Performing test 'orphaned_toast_tables'
Total runtime for test 'orphaned_toast_tables': 0:00:00.01

SUMMARY REPORT
===================================================================
Completed 14 test(s) on database 'gpadmin' at 2019-09-25 02:03:21 with elapsed time 0:00:08
Found no catalog issue

[gpadmin@sdw1 ~]$
```


So, the latest commit(code block deletion) has introduced some side effect.  Suspect some where else has used the "RangerVar"  to transfer parameters.

**Comment by gkokolatos** [comment]
> So, the latest commit(code block deletion) has introduced some side effect. Suspect some where else has used the "RangerVar" to transfer parameters.

Thank you for the reproduction. It is really helpful. I am on it.

**Comment by ginobiliwang** [comment]
> > So, the latest commit(code block deletion) has introduced some side effect. Suspect some where else has used the "RangerVar" to transfer parameters.
> 
> Thank you for the reproduction. It is really helpful. I am on it.

I think i have found the issue.

https://github.com/greenplum-db/gpdb/blob/master/src/backend/catalog/aclchk.c#L743-L763

The test case failed, because of the above code pieces. When executing on segment, no relOid could be got, since only namespace information has been transferred to segments. So, gpcheckcat output shows information only exits on master node.

**Comment by Eulerizeit** [comment]
On Wed, Sep 25, 2019 at 4:56 PM Fenggang Wang <notifications@github.com>
wrote:

> So, the latest commit(code block deletion) has introduced some side
> effect. Suspect some where else has used the "RangerVar" to transfer
> parameters.
>
> Thank you for the reproduction. It is really helpful. I am on it.
>
> I think i have found the issue.
>
>
> https://github.com/greenplum-db/gpdb/blob/master/src/backend/catalog/aclchk.c#L743-L763
>
> The test case failed, because of the above code pieces. When executing on
> segment, no relOid could be got, since only namespace information has been
> transferred to segments. So, gpcheckcat output shows information only exits
> on master node.
>

Why did this ever pass?

-- Rob

**Comment by gkokolatos** [comment]
> Why did this ever pass?
> […](#)

Please consider me adequately embarrassed.
It pass because our tests icw tests do not have full coverage and the initial code was only half right . I have updated the code and filling the tests before pushing.

Pushing shortly.

**Comment by ginobiliwang** [comment]
> On Wed, Sep 25, 2019 at 4:56 PM Fenggang Wang ***@***.***> wrote: So, the latest commit(code block deletion) has introduced some side effect. Suspect some where else has used the "RangerVar" to transfer parameters. Thank you for the reproduction. It is really helpful. I am on it. I think i have found the issue. https://github.com/greenplum-db/gpdb/blob/master/src/backend/catalog/aclchk.c#L743-L763 The test case failed, because of the above code pieces. When executing on segment, no relOid could be got, since only namespace information has been transferred to segments. So, gpcheckcat output shows information only exits on master node.
> Why did this ever pass?
> […](#)
> -- Rob


It was working because the code block below has transformed  "strVal" to "RangerVar" format on master node. Once we deleted this block, the decode part on segment could not get correct data structure.

> > > https://github.com/greenplum-db/gpdb/blob/66199b3d96ac99c60d7b5845ca0900196c148409/src/backend/catalog/aclchk.c#L473-L489

**Comment by ginobiliwang** [comment]
> It pass because our tests icw tests do not have full coverage and the initial code was only half right . I have updated the code and filling the tests before pushing.
> 
> Pushing shortly.

https://github.com/greenplum-db/gpdb/blob/66199b3d96ac99c60d7b5845ca0900196c148409/src/backend/catalog/aclchk.c#L754-L763

@gkokolatos 
This code block in function "objectNamesToOids" will perform other privileges grant tasks besides pattern "grant all on all tables in schema xxx to yyy;", that has been taken care by "objectsInSchemaToOids".   
And all of these two functions will be executed on both master and segment. 

 As a result,  do you think it is still reasonable to remove "RangerVar" ?  Since relname or relOid are required to full-fill the functionality of the above code block.

**Comment by gkokolatos** [comment]
> > It pass because our tests icw tests do not have full coverage and the initial code was only half right . I have updated the code and filling the tests before pushing.
> > Pushing shortly.
> 
> https://github.com/greenplum-db/gpdb/blob/66199b3d96ac99c60d7b5845ca0900196c148409/src/backend/catalog/aclchk.c#L754-L763
> 
> @gkokolatos
> This code block in function "objectNamesToOids" will perform other privileges grant tasks besides pattern "grant all on all tables in schema xxx to yyy;", that has been taken care by "objectsInSchemaToOids".
> And all of these two functions will be executed on both master and segment.
> 
> As a result, do you think it is still reasonable to remove "RangerVar" ? Since relname or relOid are required to full-fill the functionality of the above code block.

Please consider me adequately embarrassed from the whole mess. In the latest push the RangeVars are only expanded on the appropriate type/object combination that need to get expanded.  Please comment on https://github.com/greenplum-db/gpdb/pull/8609/commits/5875b1c2 so that I can amend if/when needed.

**Comment by ginobiliwang** [comment]
LGTM.
These conditions have clearly split up two work streams.

**Comment by gkokolatos** [comment]
Thank you all for your patience and comments. It has been merged again into master and 6X_STABLE
